File: 프로그래머스/2/12978. 배달/배달.py
import heapq

def solution(N, road, K):
    graph = [[] for _ in range(N+1)]
    dist = [float("inf")] * (N+1)
    dist[1] = 0
    
    for a,b,cost in road:
        graph[a].append((b,cost))
        graph[b].append((a,cost))
        
    heap = []
    heapq.heappush(heap, (0,1))
    while heap: 
        d,n = heapq.heappop(heap)
        for next_n, next_d in graph[n]:
            cost = d + next_d
            if cost < dist[next_n]:
                dist[next_n] = cost
                heapq.heappush(heap, (cost, next_n))
    
    return sum(1 for d in dist if d <= K)

File: 프로그래머스/2/12978. 배달/배달.js
function search(dist, arr) {
    const heap = [];
    heap.push([0, 1]);
    while (heap.length > 0) {
        heap.sort((a, b) => a[0] - b[0]);
        const [cost, node] = heap.shift();
        
        for (const [c, n] of arr[node]) {
            if (cost + c < dist[n]) {
                dist[n] = cost + c;
                heap.push([cost + c, n]);
            }
        }
    }
}

function solution(N, road, K) {
    const dist = Array(N + 1).fill(Infinity);
    dist[1] = 0;
    const arr = Array.from({ length: N + 1 }, () => []);

    for (const [u, v, w] of road) {
        arr[u].push([w, v]);
        arr[v].push([w, u]);
    }
    
    search(dist, arr);

    return dist.filter(d => d <= K).length;
}